### PR TITLE
implementing RFC 1623. This fixes #35897.

### LINKED ISSUE #35897: Tracking issue for static lifetime in statics (RFC #1623)
Tracking issue for "static lifetime in statics", rust-lang/rfcs#1623, which defaults omitted lifetimes in statics to `'static`.

cc @llogiq

### PR DESCRIPTION
**Pull request by llogiq**
This is a work in progress. In particular, I want to add more tests,
especially the compile-fail test is very bare-bones.

### CODE REVIEW

**Comment by llogiq** [comment]
Now the tests pass for me. That was too easy. :smile:

I still want to add more tests to check for possible interference with custom types (structs/enums) and aliases, but @nikomatsakis if you want to merge this, I'm fine with adding those tests in a later PR.

File: src/librustc_typeck/collect.rs [hunk 1]
```diff
@@ -1554,7 +1554,7 @@ fn type_of_def_id<'a, 'tcx>(ccx: &CrateCtxt<'a, 'tcx>,
             NodeItem(item) => {
                 match item.node {
                     ItemStatic(ref t, _, _) | ItemConst(ref t, _) => {
```
Comment (nikomatsakis) [inline src/librustc_typeck/collect.rs]: I forget, did the RFC cover both constants and statics? Presumably yes.

**Comment by nikomatsakis** [comment]
> I still want to add more tests to check for possible interference with custom types (structs/enums) and aliases, but @nikomatsakis if you want to merge this, I'm fine with adding those tests in a later PR.

Looks good to me thus far -- why not add the tests now though.

**Comment by llogiq** [comment]
Note there's still a strange mismatch between the run-pass and the compile-fail test.

File: src/test/run-pass/rfc1623.rs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+fn main() {
+    // make sure that the lifetime is actually elided (and not defaulted)
+    let x = &[1u8, 2, 3];
+    STATIC_SIMPLE_FN(x);
+    CONST_SIMPLE_FN(x);
+
+    let y = &[1u8, 2, 3];
+    STATIC_BAZ(BYTES);
+    //CONST_BAZ(y); // strangely enough, this fails
```
Comment (nikomatsakis) [inline src/test/run-pass/rfc1623.rs]: can you elaborate? how does it fail?

Comment (nikomatsakis) [inline src/test/run-pass/rfc1623.rs]: ah, I created a play version: https://is.gd/RBg85e and I see now what you mean. This is, I think, expected! The type of `Baz` would be `&'static fn(&'static [u8]) -> Option<u8>`, and yet when you supply `y` it only has a lifetime confined to the current stack frame. If you change to `CONST_BAZ(BYTES)` it should work, I think.

**Comment by llogiq** [comment]
cc @chriskrycho This will also need some documentation. I vaguely remember you wanted to be notified in such an event. I'm not sure, but perhaps a paragraph in the relevant book chapter?

**Comment by llogiq** [comment]
Error appears to be LLVM related.

**Comment by TimNN** [comment]
The LLVM error is #36077, to be fixed by #36080.

**Comment by llogiq** [comment]
Thanks, @TimNN

**Comment by chriskrycho** [comment]
👍 @llogiq – I'll review the RFC and see if I can submit a PR to the book in the next two weeks. (That's a long time because lots of travel and other commitments in the meantime, but it's on my radar now.)

**Comment by llogiq** [comment]
@chriskrycho Great, one thing I don't need to do. To save you time, the RFC basically boils down to "elide lifetimes with a `'static` default in `static`s and `const`s as per the usual elision rules. The contained tests show corner cases.

**Comment by llogiq** [comment]
@nikomatsakis now that #36080 has landed, can we bump travis? Should I make a "bump" commit?

**Comment by nikomatsakis** [comment]
@llogiq I don't know much about how travis works in that respect.

**Comment by nikomatsakis** [comment]
Er, do you just want to re-run travis results?

**Comment by nikomatsakis** [comment]
@bors r+ 

The travis failures look to be unrelated to this patch.

**Comment by petrochenkov** [comment]
@nikomatsakis 
Is it stated explicitly anywhere that this change should be insta-stable and not feature gated?
This seems reasonable and the feature gate seems to be hard to enforce, but I haven't seen this discussed at all.

File: src/test/compile-fail/rfc1623.rs.bk [hunk 3]
```diff
@@ -0,0 +1,98 @@
+// Copyright 2012 The Rust Project Developers. See the COPYRIGHT
```
Comment (rkruppe) [inline src/test/compile-fail/rfc1623.rs.bk]: Umm this seems to be a rustfmt-generated "backup" file, surely that shouldn't have been committed?

File: src/test/run-pass/rfc1623.rs.bk [hunk 4]
```diff
@@ -0,0 +1,81 @@
+// Copyright 2012 The Rust Project Developers. See the COPYRIGHT
```
Comment (rkruppe) [inline src/test/run-pass/rfc1623.rs.bk]: Same here

**Comment by nikomatsakis** [comment]
@petrochenkov argh you are correct -- that was utterly wrong on my part. We should totally feature gate this. @llogiq, can I walk you through landing that change?

**Comment by nikomatsakis** [comment]
And, yes, the `.bk` files ought to be removed. 

/me embarassed

**Comment by nikomatsakis** [comment]
(Adding the feature-gate itself is probably harder, of course, than the rest of this change, but not overly hard. The easiest way I think is to configure the rscope (or make a new one) that checks the feature gate and, when return `'static`, also issues an error. We may need to thread a span through, I have to check the signature.)

**Comment by nikomatsakis** [comment]
Nah, no big thing. :) Sharp eyes though, @petrochenkov.

**Pull request merged**